VM details (#200)
openshift/openshift-origin-design: web-console/knikubevirt/vm-details/vm-details.md

# VM Details

## Overview Tab

> ![VM-overview](img/1-0-0.jpg)

Reviewer: @cshinn Can you take a look at these charts? I believe these are using PF3 and we have updated charts in OpenShift to PF4. Also wondering if they belong above the standard detail info similar to how Nodes and Pods display charts at the top of the detail pages. 

Separate nit...should we swap status and namespace in the detail fields? Probably not important but figured it would match other resources more closely. Also we should change the project label to say namespace with the green NS badge :)

Reviewer 2: > @lizsurette This is an oldie, but do you think the % uptime popover, accessed by clicking the info icon next to Status label, is necessary? I lean toward removing it along with the info icon entirely for discoverability, utility, and ambiguity reasons.

Are you suggesting we use the link here instead like we do on the table view? I like it, if so :)

> Unfortunately I don't remember why there are two IPs included in this design. It might be worth checking with devs if these IPs are different in any way. If not then this is fine, but if they are we could probably add sub-labels like Node Details pages do:

VMs can have multiple IPs (I think there would be one per NIC?) In any case we could probably organize them better and maybe sub-label them.

Reviewer 3: @chart, I chatted with @andybraren a bit and it seems like those specific chart visuals were intended mostly as a placeholder. I expect that they would end up using the pf4 line graph like we do on the nodes pages etc. For consistency's sake, they should probably go at the top of the page as well as they are on the other pages

Reviewer 4: > Are you suggesting we use the link here instead like we do on the table view? I like it, if so :)

@lizsurette I was suggesting removing it entirely, but I suppose the [Status Popover pattern](http://openshift.github.io/openshift-origin-design/web-console/4.0-designs/status/status#status-popovers) would be a natural place to include how long the VM has been running for, and when it was last powered on/off (if that's possible and we agree worth doing).

I'm just not sure what "uptime" means in a VM context and my googling isn't very fruitful. Is our "uptime" calculation how long the VM has been "Running" for? What would take it down from 100% to 97% in this example?

Reviewer 2: > @lizsurette I was suggesting removing it entirely, but I suppose the [Status Popover pattern](http://openshift.github.io/openshift-origin-design/web-console/4.0-designs/status/status#status-popovers) would be a natural place to include how long the VM has been running for, and when it was last powered on/off (if that's possible and we agree worth doing).

Yeah I would assume we should be consistent with what we do in the table for this value.

> I'm just not sure what "uptime" means in a VM context and my googling isn't very fruitful. Is our "uptime" calculation how long the VM has been "Running" for? What would take it down from 100% to 97% in this example?

I think two values would be how long it's been running for and then how long it's been running in the last 7 days (%). I can't remember exactly where the original requirement came from though...it might be an oVirt carry over :)

Reviewer 3: Here's an example of what those charts would look like (sorry for the delay. my cluster crashed on me yesterday)
![screencapture-console-openshift-console-apps-uxd-research-devcluster-openshift-k8s-cluster-nodes-ip-10-0-133-52-ap-northeast-1-compute-internal-2019-06-06-11_14_34](https://user-images.githubusercontent.com/3300756/59044578-6252db00-884c-11e9-97cf-8293b09e1cf0.png)

The user clicks a VM’s name from the list view and is taken to that VM’s Overview tab.
The user sees detailed information about the VM’s configuration.

## YAML

Should be inline with the rest of Openshift. Can be seen [here](http://openshift.github.io/openshift-origin-design/web-console/future-openshift/code-editor-updates/code-editor-updates)

## Network Interfaces

![VM-network intefaces - add NIC](img/2-0-0.jpg)
This is Network Interfaces tab.

![VM - add NIC modal](img/2-0-1.jpg)
Default modal. The user is notified that activating this upcoming NIC requires a VM restart. 

![VM - add NIC modal - filled in](img/2-0-2.jpg)
User can choose between adding NIC only, or add NIC and restart VM to make it active.

![VM - new NIC](img/2-0-3.jpg)
The new NIC should always appear at the top of the list until the user restarts it. The new nic will have a'Pending VM restart' Link state (as a status).

![VM - New NIC - popover](img/2-0-4.jpg)
Hover on that link state (status) will display more details with a 'Restart VM action in a popover.

![VM - Pending changes notification - view changes](img/2-0-5.jpg)

> Whenever a NIC has been added or modified in a way that requires the VM to be restarted, an inline notification should appear below the tab area reminding the user to do so. This notification should persist across all tab views, including the Overview (shown previously).

Reviewer 4: ```suggestion
Whenever a NIC has been added or modified in a way that requires the VM to be restarted, an inline alert should appear below the tab area reminding the user to do so. This notification should persist across all tab views, including the Overview (shown previously).
```

> ![VM - Pending changes notification - view changes modal](img/2-0-6.jpg)

Reviewer 4: The dark backdrop behind the modal in this screen isn't fullwidth. Looks a bit strange.

Changes pending VM restart can be viewed in a dedicated modal.

> ### Edit NIC

Reviewer 4: This Edit NIC section has no written descriptive text, but all others do. Probably should, even if it seems obvious.

> ![VM - NIC item action options](img/2-1-0.jpg)

Reviewer 4: The required `Name` field isn't filled in here but the buttons to save are active. I assume the Name should be `Nic1` here.

![VM - Edit NIC modal](img/2-1-1.jpg)

![VM - Edit NIC saved changes](img/2-1-2.jpg)

## Disks

![VM - Disks tab - adding disk](img/3-0-0.jpg)

> The flow for disks is similar to NIC.

Reviewer 4: ```suggestion
```

> however, adding or editing a disk do not require VM restart.

Reviewer 4: ```suggestion
The flow for disks is similar to NICs, but adding or editing a disk does not require a VM restart.
```

![VM - add disk modal](img/3-1-0.jpg)

> The user clicks “Add Disk”, and is lead to an 'Add disk' modal.

Reviewer 4: ```suggestion
The user clicks “Add Disk”, and is led to an 'Add Disk' modal.
```

Modal title should also be capitalized (for now, until PF4).

![VM - disk list item actions](img/3-2-0.jpg)
The user edits each input and clicks the confirm button to finish. The new disk is added to the list and is sorted alphabetically.

The user can edit or delete a list item from its kebab menu for each list item.

## Snapshots

Snapshot designs can be seen [here](https://github.com/openshift/openshift-origin-design/pull/183)

## Events

TBD

## Consoles

![VM - console tab -default](img/4-0-0.jpg)
The Consoles tab allows the user to connect to the virtual machine via an in-browser or desktop-based Graphical and/or Serial console.
![VM - console dropdown options](img/4-1-0.jpg)

> The user can change the console type using the dropdown selector. 'Graphical (VNC)' is selected by default. Any currently selected option is disabled in the dropdown. 'Graphical (RDP/VNC)' open in separate windows, designated by the fa-external-link icon.

Reviewer 4: My knowledge about this is outdated, but in [Marek's video](https://youtu.be/mLeCZcSHROo?t=347) it looks like Graphical (VNC) no longer has to be a separate window. That would mean removing the external link icon here.

Author: Will go with sticking to the same window since it is the desired solution.

Two actions are available on the right-hand side. The “Open in window” action opens a separate browser window dedicated to the console view. The “Expand” action makes the console viewport fill the current window’s full width and height similar to elsewhere in OKD. The “Send Key” dropdown is disabled when disconnected from the console.

The user can start a console session using either the primary 'Connect' action button in the content area or the secondary 'Connect' button next to the console dropdown selector.

### Connecting from a powered off state

![VM - Serial (SPICE)](img/4-2-0.jpg)
If the virtual machine is powered off, the primary action button will allow the user to both start and connect to the virtual machine with one click. If the user wants to power the virtual machine on without connecting a console (an unlikely use case while in this view) they can use the actions dropdown in the top-right as usual.

> ![VM - connecting to VM](img/4-3-0.jpg)

Reviewer 4: Thank you for updating the icons around here. 👍

# VM Details

## Overview Tab

![VM-overview](img/1-0-0.jpg)
The user clicks a VM’s name from the list view and is taken to that VM’s Overview tab.
The user sees detailed information about the VM’s configuration.

## YAML

TBD - should be inline with the rest of Openshift.

## Network Interfaces

> ![VM-network intefaces - add NIC](img/2-0-0.jpg)

Author: Removing filters from NIC. Do the same for Disks?

> ![VM-network intefaces - add NIC](img/2-0-0.jpg)

Reviewer 2: Good question on Disks...I think we could make a case to go either way. For now I'm happy with the simplicity of removing it.

This is Network Interfaces tab.

> ![VM - add NIC modal](img/2-0-1.jpg)

Author: So for now, stick to title case in buttons rather the sentence case for PF4, got it.

> ![VM - add NIC modal](img/2-0-1.jpg)

Reviewer 2: Yep exactly.

> ![VM - add NIC modal](img/2-0-1.jpg)

Reviewer 4: 3 nits:

`Mac` should be capitalized to `MAC` address.

The two Save buttons should be disabled until the next screen where all required fields have been filled.

I believe OpenShift italicizes default dropdown options like "Select modal" or "Select network" to visually distinguish them slightly from fields that have been selected. Here's an example (ignore the capitalization):

![image](https://user-images.githubusercontent.com/9122899/58971155-d035ce80-8788-11e9-8d1f-c04fc8751a98.png)

> ![VM - add NIC modal](img/2-0-1.jpg)

Reviewer 5: follow up on Andy's comment about MAC being captilized, in the text "select mac address" it should be "select MAC address"

> ![VM-network intefaces - add NIC](img/2-0-0.jpg)

Reviewer 5: can we update that `button` text to be `actions`

Default modal. The user is notified that activating this upcoming NIC requires a VM restart. 

![VM - add NIC modal - filled in](img/2-0-2.jpg)
User can choose between adding NIC only, or add NIC and restart VM to make it active.

![VM - new NIC](img/2-0-3.jpg)
The new NIC should always appear at the top of the list until the user saves it. The new nic will have a'Pending VM restart' Link state (as a status).

![VM - New NIC - popover](img/2-0-4.jpg)
Hover on that link state (status) will display more details with a 'Restart VM action in a popover.

![VM - pendign changes notification - view changes](img/2-0-5.jpg)
Whenever a NIC has been added or modified in a way that requires the VM to be restarted, an inline notification should appear below the tab area reminding the user to do so. This notification should persist across all tab views, including the Overview (shown previously).

> ![VM - pendign changes notification - view changes modal](img/2-0-6.jpg)

Author: Same comment as above. use title case rather sentence case for PF4?

> ![VM - pendign changes notification - view changes modal](img/2-0-6.jpg)

Author: In the MD file right? changed

> ![VM - pendign changes notification - view changes modal](img/2-0-6.jpg)

Reviewer 2: In the header of the form section in the image.

Changes pending VM restart can be viewed in a dedictaed modal.

### Edit NIC

![VM - NIC item action options](img/2-1-0.jpg)

![VM - Edit NIC modal](img/2-1-1.jpg)

![VM - Edit NIC saved changes](img/2-1-2.jpg)

## Disks

![VM - Disks tab - adding disk](img/3-0-0.jpg)
The flow for disks is similar to NIC.
however, adding or rediting a disk do not require VM restart.

![VM - add disk modal](img/3-1-0.jpg)

> The user clicks “Add Disk”. Like adding a new NIC, the new disk with blank fields is prepended at the top of the list. The confirm button is disabled.

Author: Oh, I did make an update to the adding NIC/Disk behavior, from creating inline to a modal.
I did this cause with our recent designs with similar cases, we go with modal over inline creation.
I will update the text in the MD to align the images.

> The user clicks “Add Disk”. Like adding a new NIC, the new disk with blank fields is prepended at the top of the list. The confirm button is disabled.

Reviewer 2: Ah yes that makes more sense now why it's written that way. Makes sense to update.

The user clicks “Add Disk”, and is led to an 'Add Disk' modal.

![VM - disk list item actions](img/3-2-0.jpg)
The user edits each input and clicks the confirm button to finish. The new disk is added to the list and is sorted alphabetically.

The user can edit or delete a list item from its kebab menu for each list item.

## Snapshots

Snapshot designs can be seen [here](https://github.com/openshift/openshift-origin-design/pull/183)

## Events

TBD

## Consoles

![VM - console tab -default](img/4-0-0.jpg)
The Consoles tab allows the user to connect to the virtual machine via an in-browser or desktop-based Graphical and/or Serial console.
![VM - console dropdown options](img/4-1-0.jpg)
The user can change the console type using the dropdown selector. 'Graphical (VNC)' is selected by default. Any currently selected option is disabled in the dropdown.

> Two actions are available on the right-hand side. The “Open in window” action opens a separate browser window dedicated to the console view. The “Expand” action makes the console viewport fill the current window’s full width and height similar to elsewhere in OKD. The “Send Key” dropdown is disabled when disconnected from the console.

Reviewer 5: should this be `Open in new window`?

Author: Sounds better, I'll change that

# VM Details

## Overview Tab

![VM-overview](img/1-0-0.jpg)
The user clicks a VM’s name from the list view and is taken to that VM’s Overview tab.
The user sees detailed information about the VM’s configuration.
![VM-overview - Editing a field](img/1-0-1.jpg)

> Editing field will be aligned to the resto of the system. For now, only 'Description' and Flavour' have a pencil icon next to the field name to indicate it is editable. Hitting that pencil will opne up a modal.

Reviewer 5: ```suggestion
Editing field will be aligned to the rest of the system. For now, only 'Description' and Flavor' have a pencil icon next to the field name to indicate it is editable. Hitting that pencil will open up a modal.
```
